core/modules/replica/MessengerConnector.cc

```diff
@@ -0,0 +1,640 @@
+                             << "  _currentRequest->id=" << _currentRequest->id
+                             << " bytes=" << bytes);
+
+                        // Then receive response body into a buffer inside the wrapper
+
+                        if (syncReadMessageImpl(lock,
+                                                _currentRequest->responseBuffer,
+                                                bytes)) {
+
+                            // If something bad happened along the line then we should notify a caller about
+                            // a problem with the request, remove the request from all data structures and
+                            // restart the communication.
+
+                            request2notify = _currentRequest;
+
+                            _id2request.erase(_currentRequest->id);
+                            _currentRequest = nullptr;
+
+                            restart(lock);
+
+                        } else {
+
+                            // Finally, success!
+
+                            // Notify a caller about the successfull completion of the operation,
+                            // remove this request from all data structures and initiate a new request.
+
+                            request2notify = _currentRequest;
+                            request2notify->success = true;
+
+                            _id2request.erase(_currentRequest->id);
+                            _currentRequest = nullptr;
+
+                            sendRequest(lock);
+                        }
+                    }
+                }
+            }
+
+        } else {
+
+            // The request submission had a chance to finish (successfully or not)
+            // before the cancellation attempt was made. So, we just forget about
+            // this request and send no notification to a caller.
+            //
+            if (ec) restart(lock);
+            else    sendRequest(lock);
+        }
+    }
+
+    // Sending notifications (if requsted) outsize the lock guard
+
+    if (request2notify) request2notify->parseAndNotify();
+}
+
+boost::system::error_code MessengerConnector::syncReadFrame(util::Lock const& lock,
+                                                            replica::ProtocolBuffer& buf,
+                                                            size_t& bytes) {
+    size_t const frameLength = sizeof(uint32_t);
+    buf.resize(frameLength);
+
+    boost::system::error_code ec;
+    boost::asio::read(
+        _socket,
+        boost::asio::buffer(
+            buf.data(),
+            frameLength
+        ),
+        boost::asio::transfer_at_least(frameLength),
+        ec
+    );
+    LOGS(_log, LOG_LVL_DEBUG, context() << "syncReadFrame"
+         << "  _currentRequest->id=" << (_currentRequest ? _currentRequest->id : "")
+         << " error_code=" << ec);
+
+    if (not ec) bytes = buf.parseLength();
```

Comment:
this is triggering my 'implicit conversion to bool' bad smell alarm. I think `if (ec != boost::system::success` would be much more clear.

Reply:
I inherited this pattern from BOOST ASIO examples. And it's certainly very handy to check for error codes in:
if (ec) {...}

Reply:
Also, the error_condition class is specifically designed with two operators:
...
typedef void (*unspecified_bool_type)();
      static void unspecified_bool_true() {}

      operator unspecified_bool_type() const BOOST_SYSTEM_NOEXCEPT  // true if error
      {
        return m_val == 0 ? 0 : unspecified_bool_true;
      }
      bool operator!() const BOOST_SYSTEM_NOEXCEPT  // true if no error
      {
        return m_val == 0;
      }

Details in:
https://www.boost.org/doc/libs/1_61_0/boost/system/error_code.hpp